File: 1797-110-2116-check-if-a-parentheses-string-can-be-valid/1797-110-2116-check-if-a-parentheses-string-can-be-valid.cpp
class Solution {
public:
    bool solve(string &s, string &locked,bool forward){
        int cnt = 0 , magic = 0;
        int n =  s.size();
        for(int i = 0 ; i < n ;i++){
            int idx = forward ? i : n - i - 1;
            if(locked[idx]=='0'){
                magic++;
                continue;
            }
            if((s[idx]=='(' && forward) || (s[idx]==')' && !forward) )cnt++;
            else{
                if(cnt)cnt--;
                else if(magic)magic--;
                else return false;
            }            
        }
        return true;
    }
    bool canBeValid(string s, string locked) {
        if(s.size()&1)return 0;
        return solve(s,locked,1) & solve(s,locked,0);


    }
};

File: 1797-110-2116-check-if-a-parentheses-string-can-be-valid/1797-110-2116-check-if-a-parentheses-string-can-be-valid.java
class Solution {
    private boolean solve(String s, String locked, boolean forward) {
        int cnt = 0, magic = 0;
        int n = s.length();

        for (int i = 0; i < n; i++) {
            int idx = forward ? i : n - i - 1;

            if (locked.charAt(idx) == '0') {
                magic++;
                continue;
            }

            if ((s.charAt(idx) == '(' && forward) || (s.charAt(idx) == ')' && !forward)) {
                cnt++;
            } else {
                if (cnt > 0) {
                    cnt--;
                } else if (magic > 0) {
                    magic--;
                } else {
                    return false;
                }
            }
        }

        return true;
    }

    public boolean canBeValid(String s, String locked) {
        if (s.length() % 2 != 0) {
            return false;
        }
        return solve(s, locked, true) && solve(s, locked, false);
    }
}
